[HUDI-8920] [RFC-84] Optimize SerDe of stream records for Flink write (excluding bulk insert and append mode)

geserdugarov opened the pull request:
### Change Logs

Changes in Flink stream write into Hudi table corresponding to [Design of RFC-84, Optimized SerDe of DataStream in Flink operators](https://github.com/apache/hudi/pull/12697).

Architecture of Flink write pipelines: https://miro.com/app/board/uXjVLsCpf48=/

Changes in this PR doesn't affect bulk insert and append mode.

Main points:
- `HoodieFlinkInternalRow` is introduced. It doesn't extend `HoodieRecord` because we need to create data structure with Flink row data and Hudi metadata, constructed from Flink internal data types.
- `HoodieFlinkInternalRow` is effective for Flink processing due to implemented `HoodieFlinkInternalRowTypeInfo` and `HoodieFlinkInternalRowSerializer` with custom `serialize` and `deserialize` methods.

#### Benchmark description

`Lineitem` table from TPC-H benchmark was used. 60 mln rows, from which 20 mln rows are unique.

Data is read from 8 Kafka topics, and is written into Hudi table:
```text
'table.type'='MERGE_ON_READ',
'write.operation'='upsert',
'write.tasks'='8'
```

For simple bucket:
```text
'index.type'='BUCKET',
'hoodie.bucket.index.hash.field'='l_orderkey,l_linenumber',
'hoodie.bucket.index.num.buckets'='8',
'hoodie.index.bucket.engine'='SIMPLE'
```

For consistent hashing:
```text
'index.type'='BUCKET',
'hoodie.bucket.index.hash.field'='l_orderkey,l_linenumber',
'hoodie.index.bucket.engine'='CONSISTENT_HASHING'
```

#### Perfomance estimation results

|                                         | Previous with Kryo | `HoodieFlinkInternalRow ` | Improvement |
| --------------------------- | ------------------ | --------------------- | -------------- |
| **Non bucket**               |                            |                                 |                       |
| Data passed, GB              | 43.9                    | 29.4                          | **33.0%**      |
| Total time, s                     | 620                     | 460                          | **25.8%**      |
| **Simple bucket index**  |                            |                                |                       |
| Data passed, GB              | 19.4                     | 13.7                        | **29.4%**     |
| Total time, s                     | 350                      | 320                         | **8.6%**      |
| **Consistent hashing**   |                            |                                |                       |
| Data passed, GB              | 24.6                     | 15.8                        | **35.8%**     |
| Total time, s                     | 1040                    | 800                         | **23.1%**      |


### Impact

Flink write performance improvement.

### Risk level (write none, low medium or high below)

Medium

### Documentation Update

No need

### Contributor's checklist

- [x] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [x] Change Logs and Impact were stated clearly
- [x] Adequate tests were added if applicable
- [x] CI passed

wombatu-kun commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/configuration/FlinkOptions.java:
```diff
@@ -675,6 +675,14 @@ private FlinkOptions() {
+      .booleanType()
+      .defaultValue(false)
+      .withDescription("Optimized Flink write into Hudi table, which uses customized serialization/deserialization. "
+          + "Note, that only SIMPLE BUCKET index is supported for now.");
```
PR's title says "simple bucket **and non bucket** cases"

geserdugarov replied:
Missed it. Thanks! Fixed in 466c013e5cf486fa1fa2f7051ba1449f9858a0dc.

geserdugarov commented:
I've added IT tests in a926ed43f48fc8ef37ea56318e69ab7a0329e42d, and also manually checked restore from Flink checkpoint for non bucket case. Restoring from checkpoint was successful.

geserdugarov commented:
@danny0405 , @xiarixiaoyao, @yuzhaojing, @wombatu-kun , hi!
If you don't mind and would have time, could you, please, review this PR related to corresponding RFC https://github.com/apache/hudi/pull/12697.

Actually, the main part of proposed changes has been done in this PR. The only missed part for now is consistent hashing support (_in progress_) and bounded context (_will check it next_).

I've also finished testing.
- Run benchmark on `lineitem` table from TPC-H for supported write scenarios, and got about 30% performance increase.
- Manually checked that restore from Flink checkpoint is successful.
- Enabled `write.fast.mode` by default to run all tests in https://github.com/apache/hudi/pull/12817. There are only 2 errors in `test-flink (flink1.20, 1.11.3)`:
https://github.com/apache/hudi/actions/runs/13261652696/job/37019456652?pr=12817

```text
Error:  testScheduleSplitPlan  Time elapsed: 0.034 s  <<< ERROR!
org.apache.hudi.exception.HoodieNotSupportedException: Currently, consistent hashing is not supported with enabled 'write.fast.mode'
	at org.apache.hudi.sink.cluster.ITTestFlinkConsistentHashingClustering.prepareData(ITTestFlinkConsistentHashingClustering.java:126)
	at org.apache.hudi.sink.cluster.ITTestFlinkConsistentHashingClustering.testScheduleSplitPlan(ITTestFlinkConsistentHashingClustering.java:79)

Error:  testScheduleMergePlan  Time elapsed: 0.027 s  <<< ERROR!
org.apache.hudi.exception.HoodieNotSupportedException: Currently, consistent hashing is not supported with enabled 'write.fast.mode'
	at org.apache.hudi.sink.cluster.ITTestFlinkConsistentHashingClustering.prepareData(ITTestFlinkConsistentHashingClustering.java:126)
	at org.apache.hudi.sink.cluster.ITTestFlinkConsistentHashingClustering.testScheduleMergePlan(ITTestFlinkConsistentHashingClustering.java:104)
```
These errors are related to not supported consistent hashing yet.
All other cases are successfully passed.

geserdugarov commented:
I've squashed all commits into one, cb090d8003efa8c6703e99073bb9e33d48a952d1, and renamed `HoodieFlinkRecord` into `HoodieFlinkInternalRow` to prevent confusion, because this class doesn't extend `HoodieRecord`.

danny0405 commented:
Let's move to RFC-88 instead.

geserdugarov commented:
> Let's move to RFC-88 instead.

Ok, I will write in RFC-88.

yuzhaojing commented:
Thank you for this optimization. Given that @danny0405 and @cshuo have proposed RFC-88, let's include this optimization as a subtask within the RFC.

geserdugarov commented:
> Thank you for this optimization. Given that @danny0405 and @cshuo have proposed RFC-88, let's include this optimization as a subtask within the RFC.

Please, check this comment under RFC-88, where I've tried to described why RFC-88 and RFC-84 are different tasks:
https://github.com/apache/hudi/pull/12795#issuecomment-2655363635

We shouldn't add any abstraction for optimized serde until we get writers. Otherwise, we wouldn't reach significant performance increase.
RFC-88 is focused on processing records in writers. This RFC-84 is focused on passing data from source to writers as fast as possible.
And, I suppose, @cshuo , author of RFC-88, thinks the same:
https://github.com/apache/hudi/pull/12795#issuecomment-2641752043.

geserdugarov commented:
> Let's move to RFC-88 instead.

I've described my point for view in this comment:
https://github.com/apache/hudi/pull/12795#issuecomment-2655363635

danny0405 commented:
@geserdugarov Can we modify existing function/operators instead of adding new ones, can we also eliminate the config option, it should be always true.

geserdugarov commented:
> @geserdugarov Can we modify existing function/operators instead of adding new ones, can we also eliminate the config option, it should be always true.

Yes, we can do it. I did it this way to save previous behavior to minimize risks by migrating step by step:
- new behavior could be turned on by config, turned off by default,
- new behavior turned on by default, previous behavior is deprecated,
- previous behavior removed.

But this smooth transition is not possible for RFC-88. So, we can do optimization without additional config parameter, and changing corresponding classes without creating new ones. It would be much easier.

I will do it, and update PR as soon as possible.

geserdugarov commented:
Updated PR with optimized serde without any configuration by default. But need to fix 4 failures:
> [ERROR] Failures: 
[ERROR]   TestInputFormat.testReadBaseAndLogFilesWithDeletes:254 
Expected: is "[+I[id1, Danny, 24, 1970-01-01T00:00:00.001, par1], +I[id2, Stephen, 34, 1970-01-01T00:00:00.002, par1], -D[id3, Julian, 53, 1970-01-01T00:00:00.003, par2], +I[id4, Fabian, 31, 1970-01-01T00:00:00.004, par2], -D[id5, Sophia, 18, 1970-01-01T00:00:00.005, par3], +I[id6, Emma, 20, 1970-01-01T00:00:00.006, par3], +I[id7, Bob, 44, 1970-01-01T00:00:00.007, par4], +I[id8, Han, 56, 1970-01-01T00:00:00.008, par4], -D[id9, Jane, 19, 1970-01-01T00:00:00.006, par3]]"
     but: was "[+I[id1, Danny, 24, 1970-01-01T00:00:00.001, par1], +I[id2, Stephen, 34, 1970-01-01T00:00:00.002, par1], +I[id4, Fabian, 31, 1970-01-01T00:00:00.004, par2], +I[id6, Emma, 20, 1970-01-01T00:00:00.006, par3], +I[id7, Bob, 44, 1970-01-01T00:00:00.007, par4], +I[id8, Han, 56, 1970-01-01T00:00:00.008, par4]]"
[ERROR]   TestInputFormat.testReadChangesMergedMOR:744 
Expected: is "[-D[id1, Danny, 22, 1970-01-01T00:00:00.005, par1]]"
     but: was "[-D[id1, null, null, null, null]]"
[ERROR]   TestInputFormat.testReadChangesUnMergedMOR:773 
Expected: is "[+I[id1, Danny, 19, 1970-01-01T00:00:00.001, par1], -U[id1, Danny, 19, 1970-01-01T00:00:00.001, par1], +U[id1, Danny, 20, 1970-01-01T00:00:00.002, par1], -U[id1, Danny, 20, 1970-01-01T00:00:00.002, par1], +U[id1, Danny, 21, 1970-01-01T00:00:00.003, par1], -U[id1, Danny, 21, 1970-01-01T00:00:00.003, par1], +U[id1, Danny, 22, 1970-01-01T00:00:00.004, par1], -D[id1, Danny, 22, 1970-01-01T00:00:00.005, par1]]"
     but: was "[+I[id1, Danny, 19, 1970-01-01T00:00:00.001, par1], -U[id1, Danny, 19, 1970-01-01T00:00:00.001, par1], +U[id1, Danny, 20, 1970-01-01T00:00:00.002, par1], -U[id1, Danny, 20, 1970-01-01T00:00:00.002, par1], +U[id1, Danny, 21, 1970-01-01T00:00:00.003, par1], -U[id1, Danny, 21, 1970-01-01T00:00:00.003, par1], +U[id1, Danny, 22, 1970-01-01T00:00:00.004, par1]]"
[ERROR]   TestInputFormat.testReadWithDeletesMOR:318 
Expected: is "[+I[id1, Danny, 24, 1970-01-01T00:00:00.001, par1], +I[id2, Stephen, 34, 1970-01-01T00:00:00.002, par1], -D[id3, Julian, 53, 1970-01-01T00:00:00.003, par2], -D[id5, Sophia, 18, 1970-01-01T00:00:00.005, par3], -D[id9, Jane, 19, 1970-01-01T00:00:00.006, par3]]"
     but: was "[+I[id1, Danny, 24, 1970-01-01T00:00:00.001, par1], +I[id2, Stephen, 34, 1970-01-01T00:00:00.002, par1], -D[id3, null, null, null, null], -D[id5, null, null, null, null], -D[id9, null, null, null, null]]"
[INFO] 
[ERROR] Tests run: 362, Failures: 4, Errors: 0, Skipped: 0

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/bucket/BucketStreamWriteFunction.java:
```diff
@@ -121,21 +126,34 @@ public void snapshotState() {
+                             ProcessFunction<HoodieFlinkInternalRow, Object>.Context context,
+                             Collector<Object> collector) throws Exception {
+    RowData row = record.getRowData();
+    // [HUDI-8969] Analyze how to get rid of excessive conversions, should be a subtask of RFC-88
```
Maybe in this PR we change it to buffer RowData instead, and convert it to `HoodieRecord` before flushing.

geserdugarov replied:
Need some time to think about it. Actually, it would be good move towards our common goal.

geserdugarov replied:
Good proposal. Did it in
https://github.com/apache/hudi/pull/12796/commits/272b9298e575ed49caf39d58428687d716129366#diff-60f2e3e9c71c40267ce166066fe216564fe75bce2ce97d2a0c2a22ec31dc2983R475

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/bucket/ConsistentBucketStreamWriteFunction.java:
```diff
@@ -53,15 +52,16 @@ public class ConsistentBucketStreamWriteFunction<I> extends StreamWriteFunction<
     super.open(parameters);
     List<String> indexKeyFields = Arrays.asList(config.getString(FlinkOptions.INDEX_KEY_FIELD).split(","));
     this.updateStrategy = new FlinkConsistentBucketUpdateStrategy(this.writeClient, indexKeyFields);
+    preparePayload();
```
why this change?

geserdugarov replied:
`ConsistentBucketStreamWriteFunction::open` overrides `StreamWriteFunction::open`, but uses `StreamWriteFunction::processElement`. And in `StreamWriteFunction::processElement` `payloadCreation` should be already initialized.

But if I will find a way to buffer `RowData` instead of `HoodieRecord` without breaking everything, then possibly there would be no need in this initialization.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/partitioner/BucketAssignFunction.java:
```diff
@@ -81,8 +79,12 @@ public class BucketAssignFunction<K, I, O extends HoodieRecord<?>>
+   * Otherwise, for any chosen data structure we should implement custom serializer.
    */
-  private ValueState<HoodieRecordGlobalLocation> indexState;
+  private ValueState<Tuple2<StringData, StringData>> indexState;
```
This breaks the compatibility of the state.

geserdugarov replied:
My bad, this place is actually doesn't require optimization because here serde will be executed only during checkpoints, which is negligible in comparison of serde cost of each stream record.

I've rollbacked previous implementation of `indexState`, which contains `HoodieRecordGlobalLocation`.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/partitioner/BucketAssignFunction.java:
```diff
@@ -154,88 +147,86 @@ public void initializeState(FunctionInitializationContext context) {
+        String partitionFromState = indexStateValue.getField(0).toString();
+        String fileIdFromState = indexStateValue.getField(1).toString();
+        if (!Objects.equals(partitionFromState, partition)) {
+          // [HUDI-8996] No delete records for Flink upsert if partition path changed
```
Fix it in this PR.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/transform/RowDataToHoodieFunction.java:
```diff
@@ -89,29 +79,17 @@ public void open(Configuration parameters) throws Exception {
-  @SuppressWarnings("rawtypes")
-  private HoodieRecord toHoodieRecord(I record) throws Exception {
+  public O map(I record) throws Exception {
+    // [HUDI-8969] Analyze how to get rid of excessive conversions, should be a subtask for RFC-88
```
The avro record is used to get the record key and partition path, just get it with RowDataKeyGen should be fine.

geserdugarov replied:
I propose to leave this part as it is in this PR.

There are two main problem with `RowDataKeyGen`:
1. it doesn't support `CustomAvroKeyGenerator`,
2. it doesn't allow to process case, when partition column is string, and we set key generator as:
```text
'hoodie.datasource.write.keygenerator.class'='org.apache.hudi.keygen.TimestampBasedAvroKeyGenerator',
'hoodie.keygen.timebased.timestamp.type' = 'DATE_STRING',
'hoodie.keygen.timebased.input.dateformat' = 'yyyy-MM-dd HH:mm:ss',
'hoodie.keygen.timebased.output.dateformat' = 'yyyy-MM-dd'
```

Maybe those scenarios are not used by users. But I'm afraid that switch to `RowDataKeyGen` could break some specific use cases.

I propose to review key generators in a separate task, [HUDI-9042](https://issues.apache.org/jira/browse/HUDI-9042). This task should be also useful for RFC-88. And this solution will localize any related changes in smaller commit, which could be easily reverted.

danny0405 replied:
Let's just switch to `RowDataKeyGen` and resolve the corner cases in separate PR, `CustomAvroKeyGenerator` does not even work for master branch actually.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/sink/StreamWriteFunction.java:
```diff
@@ -140,8 +172,29 @@ public void snapshotState() {
+  public void processElement(HoodieFlinkInternalRow record,
+                             ProcessFunction<HoodieFlinkInternalRow, Object>.Context ctx,
+                             Collector<Object> out) throws Exception {
+    RowData row = record.getRowData();
```
Can we buffer rowdata instead and do the conversion right before flushing? Can we abstract the conversion as a utility method.

geserdugarov replied:
Postponed conversion until the moment before flushing. And in the latest version conversion is localized in a separate method `StreamWriteFunction::convertToHoodieRecords`.

geserdugarov commented:
Pushed update from 1st part of review to check current state of all other functionality.

But need some time to work on two remained review comments about buffering of `RowData` and using of `RowDataKeyGen`.

geserdugarov commented:
**PR is not ready for merge**. Need to finish work on two remained major review comments about buffering of RowData and using of RowDataKeyGen.
And also need to implement manual testing of the latest version.

geserdugarov commented:
@danny0405 , I've switched `RowDataToHoodieFunction` to use of `RowDataKeyGen`.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/bucket/ITTestBucketStreamWrite.java:
```diff
@@ -78,10 +84,10 @@ public void testBucketStreamWriteAfterRollbackFirstFileGroupCreation(boolean isC
+    doWrite(tablePath, isCow, 1, "upsert");
+    doDeleteCommit(tablePath);
+    doWrite(tablePath, isCow, 1, "upsert");
+    doWrite(tablePath, isCow, 1, "upsert");
```
bucket index is always "upsert", do we have other options?

geserdugarov replied:
We can also use `bulk_insert`. But I agree, that it adds confusing. Removed it in 64a2c129f64c0402a784073dae1071c7cf5f0029.

danny0405 commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/bucket/ITTestBucketStreamWrite.java:
```diff
@@ -91,7 +97,20 @@ public void testBucketStreamWriteAfterRollbackFirstFileGroupCreation(boolean isC
+    String tablePath = tempFile.getAbsolutePath();
+    doWrite(tablePath, isCow, 2, "upsert");
+    if (isCow) {
+      TestData.checkWrittenDataCOW(tempFile, EXPECTED_LIST);
```
why test COW with different data list?

geserdugarov replied:
`TestData::checkWrittenData` expects only one base file:
https://github.com/apache/hudi/blob/3525983c90eb2fd02682596ada304ba11897290e/hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/utils/TestData.java#L742-L744
So I used `TestData::checkWrittenDataCOW` instead.

But `TestData::checkWrittenDataCOW` expects data list for comparison as `Map<String, List<String>>`. So I made a duplicate for `Map<String, String> EXPECTED` and called it `EXPECTED_LIST`. But I am convinced, that I should clean this part after your comment.

In 64a2c129f64c0402a784073dae1071c7cf5f0029, I left only `Map<String, List<String>> EXPECTED_AS_LIST` and generate corresponding `Map<String, String> EXPECTED` at this part:
https://github.com/apache/hudi/pull/12796/commits/64a2c129f64c0402a784073dae1071c7cf5f0029#diff-5e48b1082ac20d788613b6b766aaa7cd80ebaff79ebba86da2f4c9d4176ae1fcR73-R76

danny0405 replied:
> TestData::checkWrittenData expects only one base file

So why there are multiple base files with this patch, Hudi has small file profiling and only 1 latest base file is expected.

geserdugarov replied:
`TestData::checkWrittenData` doesn't expect data with buckets, when we have multiple base files in one partition. So `ITTestBucketStreamWrite::testBucketStreamWriteAfterRollbackFirstFileGroupCreation` works only because `bucketNum` is set to 1. If we will change it to 2, then this test will fail:
https://github.com/apache/hudi/blob/ec044790fe71f04710472ad63e7859f1b29e3985/hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/bucket/ITTestBucketStreamWrite.java#L77-L92

I suppose we need to fix it with revisiting of `TestData::checkWrittenData`, `TestData::checkWrittenDataCOW`, `TestData::checkWrittenDataMOR` to prevent such mistakes in future. I could work on it later. I've created a corresponding task [HUDI-9056](https://issues.apache.org/jira/browse/HUDI-9056).

geserdugarov commented:
@danny0405 , I've made changes in `ITTestBucketStreamWrite` based on your comments.

yuruguo commented on hudi-flink-datasource/hudi-flink/src/test/java/org/apache/hudi/sink/ITTestDataStreamWrite.java:
```diff
@@ -190,40 +182,71 @@ private void testWriteCopyOnWriteWithClustering(boolean sortClusteringEnabled) t
       String jobName,
       int checkpoints,
       boolean restartJob,
+      boolean useFastMode,
```
`useFastMode` is unuseful

cshuo commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/model/HoodieFlinkInternalRowSerializer.java:
```diff
@@ -0,0 +1,169 @@
+    HoodieFlinkInternalRow record;
+    if (!isIndexRecord) {
+      RowData rowData = rowDataSerializer.deserialize(source);
+      record = new HoodieFlinkInternalRow(
```
One thing needs to note here is the `StringData` fields deserialized from `source` is `BinaryStringData` which is bytes based, and calling `toString` will materialize the internal bytes into Java String object, which may not be necessary and incurs extra costs. We can achieve 'lazy materialization' by adding a constructor to accept 'StringData' parameters directly.

geserdugarov replied:
Good point, but possibly it's not suitable here. I propose to research this question in a separate task, because it would be a tough one. But quick check of these changes led me to increased time for non bucket case from 460 s up to 590 s.

cshuo replied:
For performance lost in this case, I am wondering whether the benchmark metric is stable for multiple runs even with same settings. The performance gain by lazy materializing string bytes maybe not obvious in scenario of small data, but it'll not lead to performance degradation instead.

geserdugarov replied:
Rechecked using constructors of `HoodieFlinkInternalRow` with `StringData`, but still get worse performance for these changes. I propose to research it under [HUDI-9055](https://issues.apache.org/jira/browse/HUDI-9055).

geserdugarov commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/model/HoodieFlinkInternalRowTypeInfo.java:
```diff
@@ -0,0 +1,107 @@
+
+  private final RowType rowType;
+
+  private final TypeInformation<RowData> rowDataInfo;
```
`rowDataInfo` is used in `HoodieFlinkInternalRowTypeInfo::getTotalFields`, which should return number of fields including nested ones. But I tried to return `HoodieFlinkInternalRowTypeInfo::getTotalFields` as arity, but then write time for non bucket case increased from 460 s up to 520 s, which is bad.

cshuo replied:
`getArity` and `getTotalFields` are used in Flink internal `CompositeType`, for eg, `TupleTypeInfoBase` or `PojoTypeInfo`, and are useless for other cases. Actually, we can simple return 1 for both cases, just like what `InternalTypeInfo` for RowData does or you can refer to other system, e.g, Iceberg/Paimon, which also have some customized TypeInfo & Serializer. 

For the performance lost, it's weird, because I believe`getArity` and `getTotalFields`will not be called at all. I wondering whether the benchmark metric is stable for multiple runs even with same settings.

geserdugarov replied:
Recheck everything again. With `rowDataInfo` behavior is as expected. Removed it in 70904b13a204e9ea917e73514a82d018bb30ec3d.

geserdugarov commented:
@cshuo , thanks for your review. I understand your comments, and want to research proposed changes in a separate task. It would require more profiling, and, possibly, analysis of heap dumps.

For now, I've implemented quick test of proposed changes, but in both cases I lost a huge part of gained performance improvement. For non bucket case, total write time increased from 460 s to 520 s and 590 s correspondingly.

And mentioned research is needed to have an answer, why we lost performance instead of improving it with proposed changes, which is curious behavior.

geserdugarov commented:
> @geserdugarov The changes I propose is for trying to align the ser/de performance with the Flink internal `RowDataSerializer` for RowData. And for the performance lost, I suggest you can check the benchmark stability firstly, and make sure there is no extra variables.

Rechecked your proposed changes, and removed `rowDataInfo` in 70904b13a204e9ea917e73514a82d018bb30ec3d. But implementation of `HoodieFlinkInternalRow` constructors with `StringData`, but still lead to worse performance. I propose to research it under [HUDI-9055](https://issues.apache.org/jira/browse/HUDI-9055).

cshuo commented:
> Rechecked your proposed changes, and removed `rowDataInfo` in [70904b1](https://github.com/apache/hudi/commit/70904b13a204e9ea917e73514a82d018bb30ec3d). Behavior is as expected. But implementation of `HoodieFlinkInternalRow` constructors with `StringData` for serde still leads to worse performance. I propose to research it under [HUDI-9055](https://issues.apache.org/jira/browse/HUDI-9055).

Thanks for checking. Since the 'lazy materialization' is a "best to have" feature, I'm ok with researching it in another JIRA and pick it up later.

geserdugarov commented:
@hudi-bot run azure

geserdugarov commented:
Something wrong with CI. Azure job didn't start on 70904b13a204e9ea917e73514a82d018bb30ec3d.
And `hudi-bot run azure` triggered Azure job on the previous commit:
https://dev.azure.com/apachehudi/hudi-oss-ci/_build/results?buildId=3735&view=results

I will try to amend commit then.

geserdugarov commented:
No way. `hudi-bot` lost results of already finished Azure CI on squashed commit 66727fa8ec0cc68d4e810c5268439f179de6ff7f:
https://dev.azure.com/apachehudi/hudi-oss-ci/_build/results?buildId=3736&view=results

![azure_66727fa8](https://github.com/user-attachments/assets/7f6b6a5e-c888-40b1-8d6f-01704b56f126)

geserdugarov commented:
@danny0405 , should I restart CI again, or I could leave it in this state?

danny0405 commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/model/HoodieFlinkInternalRowSerializer.java:
```diff
@@ -0,0 +1,169 @@
+
+  @Override
+  public HoodieFlinkInternalRow createInstance() {
+    throw new UnsupportedOperationException();
```
Are we clear about the behavior already, is there any case this exception can throw?

geserdugarov replied:
Added clarifying message in ef000ad30a954dbe506d42ad88a116d48dc484e0.

danny0405 commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/model/HoodieFlinkInternalRowSerializer.java:
```diff
@@ -0,0 +1,169 @@
+
+  @Override
+  public HoodieFlinkInternalRow copy(HoodieFlinkInternalRow from, HoodieFlinkInternalRow reuse) {
+    throw new UnsupportedOperationException();
```
Are we clear about the behavior already, is there any case this exception can throw?

geserdugarov replied:
Here we should return `HoodieFlinkInternalRow reuse`, but in `HoodieFlinkInternalRow` `recordKey` and `partitionPath` are final. So, we don't allow to do this type of `copy`.

And from Flink side this behavior is expected:
> The method makes an attempt to store the copy in the given reuse element, if the type is mutable. This is, however, not guaranteed.

https://github.com/apache/flink/blob/c92f208f4a074812b5314f2b4858f244b5ecd9fe/flink-core/src/main/java/org/apache/flink/api/common/typeutils/TypeSerializer.java#L103-L110

geserdugarov replied:
Added clarifying message in ef000ad30a954dbe506d42ad88a116d48dc484e0.

danny0405 commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/model/HoodieFlinkInternalRowSerializer.java:
```diff
@@ -0,0 +1,169 @@
+
+  @Override
+  public int getLength() {
+    return -1;
```
Are we clear about the effect of the hard code -1?

geserdugarov replied:
Yes, here `-1` is expected for variable length data types:
https://github.com/apache/flink/blob/c92f208f4a074812b5314f2b4858f244b5ecd9fe/flink-core/src/main/java/org/apache/flink/api/common/typeutils/TypeSerializer.java#L115-L119

geserdugarov replied:
Added clarifying comment in ef000ad30a954dbe506d42ad88a116d48dc484e0.

cshuo commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/model/HoodieFlinkInternalRow.java:
```diff
@@ -0,0 +1,134 @@
+        this.instantTime.toString(),
+        this.operationType.toString(),
+        this.isIndexRecord.getValue(),
+        this.rowData);
```
rowData here is not deep copied, and HoodieFlinkInternalRow#copy() is used in HoodieFlinkInternalRowSerializer#copy(), which requires deep copy.

geserdugarov replied:
Thanks for catching it! Fixed in ef000ad30a954dbe506d42ad88a116d48dc484e0.

geserdugarov commented:
Rechecked that all writes, non bucket, simple bucket, and consistent hashing, are not broken after ef000ad30a954dbe506d42ad88a116d48dc484e0.

cshuo commented on hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/client/model/HoodieFlinkInternalRow.java:
```diff
@@ -0,0 +1,137 @@
+    return rowData;
+  }
+
+  public HoodieFlinkInternalRow copy(RowType rowType) {
```
Better to pass `RowDataSerializer` to `copy()`, to avoid creating serializer for each record.

geserdugarov replied:
You're right. My bad. Fixed by e14cb6a559fb65152a29c5c78a90f006458cc7a0.

geserdugarov commented:
Azure CI is successful on 8dd7d65563b0d2565f426aff59d98cbc5071e9bc:
https://dev.azure.com/apachehudi/hudi-oss-ci/_build/results?buildId=3784&view=results

![azure_ci](https://github.com/user-attachments/assets/2458b3ae-f99a-46c2-8fe0-e0ec5a97f723)

geserdugarov commented:
@hudi-bot run azure

danny0405 reviewed (approved):
I have no more comments now, cc @cshuo for a final review.

cshuo commented:
LGTM, +1. cc @danny0405

geserdugarov commented:
I've just added one small commit 8b4ad61d5dc615bb609e1c7e0c1743656741dfe8, which rewrites conversion into list of `HoodieRecord`s using for loop instead of Stream API for small performance improvement.

vinothchandar commented:
Great hustle, everyone!

danny0405 commented:
@geserdugarov Thanks for the great work, please investigate and research the left over issues.

geserdugarov commented:
> @geserdugarov Thanks for the great work, please investigate and research the left over issues.

Thanks. Yes, I've started to work on them.
